Meson port

Salamandar opened the pull request:
Here's a fully working Meson build for Nemo.
The full build is now 3× faster than with Autotools… Pretty cool :)
Hope you're okay with merging this.

Salamandar commented:
I renamed the icons because I could not find how to rename the files easily at install. It's also cleaner in the src directory.

JosephMcc commented:
I'm curious where you are testing this. I didn't look through it carefully and and have only limited experience with meson. Mint currently has a fairly old version of meson/ninja that doesn't support some of the newer functionality. Also, how does it work with the debian build system? I've never actually tried it.

Salamandar commented:
Oh, I'm trying it on ArchLinux (well Manjaro). 
I really don't know how to set up a debian build system. Maybe I'll check it out.
Meson is pretty simple (compared to autotools) and I just used the same structure/tweaks than GStreamer.
BTW, this doesn't break Autotools so the "stable debian packages" can still use it.

leigh123linux commented:
docs/reference/libnemo-extension/meson.build is missing

```
[leigh@localhost build]$ meson
The Meson build system
Version: 0.40.1
Source dir: /home/leigh/Downloads/nemo-MesonPort
Build dir: /home/leigh/Downloads/nemo-MesonPort/build
Build type: native build
Project name: nemo
Native c compiler: cc (gcc 7.1.1)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Has header "malloc.h": YES
Has header "sys/mount.h": YES
Has header "sys/param.h": YES
Has header "sys/vfs.h": YES
Has header "X11/XF86keysym.h": YES
Checking for function "mallopt": YES
Found pkg-config: /usr/bin/pkg-config (1.3.6)
Native dependency libexif found: YES 0.6.21
Native dependency exempi-2.0 found: YES 2.4.2
Native dependency tracker-sparql-1.0 found: YES 1.12.0
Program update-mime-database found: YES (/usr/bin/update-mime-database)
Program perl5 found: NO
Program perl found: YES (/usr/bin/perl)
Program glib-genmarshal found: YES (/usr/bin/glib-genmarshal)
Library m found: YES
Native dependency glib-2.0 found: YES 2.52.2
Native dependency gtk+-3.0 found: YES 3.22.15
Native dependency cinnamon-desktop found: YES 3.4.0
Native dependency gail-3.0 found: YES 3.22.15
Native dependency libxml-2.0 found: YES 2.9.4
Native dependency x11 found: YES 1.6.5
Native dependency gthread-2.0 found: YES 2.52.2
Native dependency gmodule-no-export-2.0 found: YES 2.52.2
Native dependency gio-2.0 found: YES 2.52.2
Native dependency gobject-2.0 found: YES 2.52.2
Native dependency gobject-introspection-1.0 found: YES 1.52.1
Native dependency gio-unix-2.0 found: YES 2.52.2
Native dependency libnotify found: YES 0.7.7
Configuring nemo-autorun-software.desktop using configuration
Configuring nemo-autostart.desktop using configuration
Configuring nemo.desktop using configuration
Configuring org.Nemo.service using configuration
Configuring org.nemo.freedesktop.FileManager1.service using configuration
Configuring nemo.xml using configuration
Configuring org.nemo.root.policy using configuration
Program g-ir-scanner found: YES (/usr/bin/g-ir-scanner)
Program g-ir-compiler found: YES (/usr/bin/g-ir-compiler)
Native dependency gobject-introspection-1.0 found: YES 1.52.1
Configuring libnemo-extension.pc using configuration
Program gdbus-codegen found: YES (/usr/bin/gdbus-codegen)
Native dependency glib-2.0 found: YES 2.52.2
Configuring org.nemo.gschema.xml using configuration
Program gtkdoc-scan found: YES (/usr/bin/gtkdoc-scan)

Meson encountered an error in file docs/reference/meson.build, line 1, column 0:
Nonexistent build def file docs/reference/libnemo-extension/meson.build.
```

Salamandar commented:
Added. The directory was in the .gitignore ^^

leigh123linux commented:
This PR  FTBFS if tracker is disabled

```
$ meson -Dselinux=true -Denable-tracker=false
The Meson build system
Version: 0.40.1
Source dir: /home/leigh/Downloads/nemo-MesonPort
Build dir: /home/leigh/Downloads/nemo-MesonPort/build
Build type: native build
Project name: nemo
Native c compiler: cc (gcc 7.1.1)
Build machine cpu family: x86_64
Build machine cpu: x86_64
Has header "malloc.h": YES
Has header "sys/mount.h": YES
Has header "sys/param.h": YES
Has header "sys/vfs.h": YES
Has header "X11/XF86keysym.h": YES
Checking for function "mallopt": YES
Found pkg-config: /usr/bin/pkg-config (1.3.6)
Native dependency libexif found: YES 0.6.21
Native dependency exempi-2.0 found: YES 2.4.2
Native dependency libselinux found: YES 2.6
Program update-mime-database found: YES (/usr/bin/update-mime-database)
Program perl5 found: NO
Program perl found: YES (/usr/bin/perl)
Program glib-genmarshal found: YES (/usr/bin/glib-genmarshal)
Library m found: YES
Native dependency glib-2.0 found: YES 2.52.2
Native dependency gtk+-3.0 found: YES 3.22.15
Native dependency cinnamon-desktop found: YES 3.4.0
Native dependency gail-3.0 found: YES 3.22.15
Native dependency libxml-2.0 found: YES 2.9.4
Native dependency x11 found: YES 1.6.5
Native dependency gthread-2.0 found: YES 2.52.2
Native dependency gmodule-no-export-2.0 found: YES 2.52.2
Native dependency gio-2.0 found: YES 2.52.2
Native dependency gobject-2.0 found: YES 2.52.2
Native dependency gobject-introspection-1.0 found: YES 1.52.1
Native dependency gio-unix-2.0 found: YES 2.52.2
Native dependency libnotify found: YES 0.7.7
Configuring nemo-autorun-software.desktop using configuration
Configuring nemo-autostart.desktop using configuration
Configuring nemo.desktop using configuration
Configuring org.Nemo.service using configuration
Configuring org.nemo.freedesktop.FileManager1.service using configuration
Configuring nemo.xml using configuration
Configuring org.nemo.root.policy using configuration
Program g-ir-scanner found: YES (/usr/bin/g-ir-scanner)
Program g-ir-compiler found: YES (/usr/bin/g-ir-compiler)
Native dependency gobject-introspection-1.0 found: YES 1.52.1
Configuring libnemo-extension.pc using configuration
Program gdbus-codegen found: YES (/usr/bin/gdbus-codegen)
Native dependency glib-2.0 found: YES 2.52.2
Configuring org.nemo.gschema.xml using configuration
Program gtkdoc-scan found: YES (/usr/bin/gtkdoc-scan)
Configuring version.xml using configuration
Configuring config.h using configuration
Message: nemo-3.4.1

    prefix:                 /usr/local
    source code location:   /home/leigh/Downloads/nemo-MesonPort
    compiler:           gcc
    debugging support:  debug

    libexif support:    true
    exempi  support:    true
    Tracker support:    false

    nemo-extension documentation: true
    nemo-extension introspection: introspection ?
Build targets in project: 25
[leigh@localhost build]$ ninja
[19/266] Compiling c object 'eel/eel@sta/eel-editable-label.c.o'
../eel/eel-editable-label.c: In function ‘eel_editable_label_get_block_cursor_location’:
../eel/eel-editable-label.c:1325:3: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
   if (layout_line == NULL)
   ^~
../eel/eel-editable-label.c:1328:5: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘if’
     text = pango_layout_get_text (label->layout);
     ^~~~
[54/266] Compiling c object 'libnemo-p...te/nemo-private@sha/nemo-bookmark.c.o'
../libnemo-private/nemo-bookmark.c: In function ‘nemo_bookmark_new’:
../libnemo-private/nemo-bookmark.c:703:5: warning: this ‘else’ clause does not guard... [-Wmisleading-indentation]
     else
     ^~~~
../libnemo-private/nemo-bookmark.c:706:2: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘else’
  new_bookmark = NEMO_BOOKMARK (g_object_new (NEMO_TYPE_BOOKMARK,
  ^~~~~~~~~~~~
[59/266] Compiling c object 'libnemo-p...private@sha/nemo-desktop-metadata.c.o'
../libnemo-private/nemo-desktop-metadata.c: In function ‘nemo_desktop_metadata_init’:
../libnemo-private/nemo-desktop-metadata.c:143:15: warning: variable ‘file’ set but not used [-Wunused-but-set-variable]
     GKeyFile *file;
               ^~~~
../libnemo-private/nemo-desktop-metadata.c: In function ‘nemo_desktop_get_metadata_stringv’:
../libnemo-private/nemo-desktop-metadata.c:252:43: warning: passing argument 4 of ‘g_key_file_get_string_list’ from incompatible pointer type [-Wincompatible-pointer-types]
                                           &length,
                                           ^
In file included from /usr/include/glib-2.0/glib.h:55:0,
                 from ../libnemo-private/nemo-desktop-metadata.h:28,
                 from ../libnemo-private/nemo-desktop-metadata.c:27:
/usr/include/glib-2.0/glib/gkeyfile.h:206:11: note: expected ‘gsize * {aka long unsigned int *}’ but argument is of type ‘guint * {aka unsigned int *}’
 gchar   **g_key_file_get_string_list        (GKeyFile             *key_file,
           ^~~~~~~~~~~~~~~~~~~~~~~~~~
[60/266] Compiling c object 'libnemo-p...mo-private@sha/nemo-desktop-utils.c.o'
../libnemo-private/nemo-desktop-utils.c: In function ‘nemo_desktop_utils_get_primary_monitor’:
../libnemo-private/nemo-desktop-utils.c:100:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
[80/266] Compiling c object 'libnemo-p...e/nemo-private@sha/nemo-job-queue.c.o'
../libnemo-private/nemo-job-queue.c: In function ‘nemo_job_queue_add_new_job’:
../libnemo-private/nemo-job-queue.c:208:5: warning: this ‘else’ clause does not guard... [-Wmisleading-indentation]
     else
     ^~~~
../libnemo-private/nemo-job-queue.c:211:2: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘else’
  g_signal_emit (self, signals[NEW_JOB], 0, NULL);
  ^~~~~~~~~~~~~
[81/266] Compiling c object 'libnemo-private/nemo-private@sha/nemo-file.c.o'
In file included from /usr/include/glib-2.0/gobject/gobject.h:24:0,
                 from /usr/include/glib-2.0/gobject/gbinding.h:29,
                 from /usr/include/glib-2.0/glib-object.h:23,
                 from /usr/include/glib-2.0/gio/gioenums.h:28,
                 from /usr/include/glib-2.0/gio/giotypes.h:28,
                 from /usr/include/glib-2.0/gio/gio.h:26,
                 from /usr/include/gtk-3.0/gdk/gdkapplaunchcontext.h:28,
                 from /usr/include/gtk-3.0/gdk/gdk.h:32,
                 from /usr/include/gtk-3.0/gtk/gtk.h:30,
                 from ../libnemo-private/nemo-file.h:28,
                 from ../libnemo-private/nemo-file.c:26:
../libnemo-private/nemo-file.c: In function ‘nemo_file_get_metadata_list’:
/usr/include/glib-2.0/gobject/gtype.h:2221:42: warning: return from incompatible pointer type [-Wincompatible-pointer-types]
 #define _G_TYPE_IGC(ip, gt, ct)         ((ct*) (((GTypeInstance*) ip)->g_class))
/usr/include/glib-2.0/gobject/gtype.h:525:66: note: in expansion of macro ‘_G_TYPE_IGC’
 #define G_TYPE_INSTANCE_GET_CLASS(instance, g_type, c_type)     (_G_TYPE_IGC ((instance), (g_type), c_type))
                                                                  ^~~~~~~~~~~
../libnemo-private/nemo-file.h:54:4: note: in expansion of macro ‘G_TYPE_INSTANCE_GET_CLASS’
   (G_TYPE_INSTANCE_GET_CLASS ((obj), NEMO_TYPE_FILE, NemoFileClass))
    ^~~~~~~~~~~~~~~~~~~~~~~~~
../libnemo-private/nemo-file.c:3548:16: note: in expansion of macro ‘NEMO_FILE_GET_CLASS’
         return NEMO_FILE_GET_CLASS (file)->get_metadata_as_list (file, key);
                ^~~~~~~~~~~~~~~~~~~
../libnemo-private/nemo-file.c: In function ‘nemo_file_should_show_thumbnail’:
../libnemo-private/nemo-file.c:4107:5: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
     if (file->details->thumbnail_access_problem)
     ^~
../libnemo-private/nemo-file.c:4110:2: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘if’
  if (show_image_thumbs == NEMO_SPEED_TRADEOFF_ALWAYS) {
  ^~
../libnemo-private/nemo-file.c: In function ‘nemo_file_construct_tooltip’:
../libnemo-private/nemo-file.c:7497:55: warning: pointer targets in passing argument 2 of ‘nemo_file_get_directory_item_count’ differ in signedness [-Wpointer-sign]
         if (nemo_file_get_directory_item_count (file, &item_count, NULL)) {
                                                       ^
../libnemo-private/nemo-file.c:4811:1: note: expected ‘guint * {aka unsigned int *}’ but argument is of type ‘gint * {aka int *}’
 nemo_file_get_directory_item_count (NemoFile *file,
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[91/266] Compiling c object 'libnemo-p...sha/nemo-mime-application-chooser.c.o'
../libnemo-private/nemo-mime-application-chooser.c: In function ‘application_selected_cb’:
../libnemo-private/nemo-mime-application-chooser.c:238:5: warning: this ‘if’ clause does not guard... [-Wmisleading-indentation]
     if (chooser->details->dialog_ok)
     ^~
../libnemo-private/nemo-mime-application-chooser.c:241:2: note: ...this statement, but the latter is misleadingly indented as if it were guarded by the ‘if’
  g_object_unref (default_app);
  ^~~~~~~~~~~~~~
[94/266] 'Generating Nemo-3.0.gir with a custom command.'
<unknown>:: Warning: Nemo: (Interface)FileInfo: Couldn't find associated structure for 'FileInfo'
[98/266] Compiling c object 'libnemo-p...ate@sha/nemo-search-engine-simple.c.o'
In file included from /usr/lib64/glib-2.0/include/glibconfig.h:9:0,
                 from /usr/include/glib-2.0/glib/gtypes.h:32,
                 from /usr/include/glib-2.0/glib/galloca.h:32,
                 from /usr/include/glib-2.0/glib.h:30,
                 from /usr/include/glib-2.0/gobject/gbinding.h:28,
                 from /usr/include/glib-2.0/glib-object.h:23,
                 from ../libnemo-private/nemo-search-engine.h:27,
                 from ../libnemo-private/nemo-search-engine-simple.h:27,
                 from ../libnemo-private/nemo-search-engine-simple.c:25:
../libnemo-private/nemo-search-engine-simple.c: In function ‘strsplit_esc_n’:
../libnemo-private/nemo-search-engine-simple.c:95:34: warning: comparison between pointer and integer
  g_return_val_if_fail (delimiter != NULL, NULL);
                                  ^
/usr/include/glib-2.0/glib/gmacros.h:379:25: note: in definition of macro ‘G_LIKELY’
 #define G_LIKELY(expr) (expr)
                         ^~~~
../libnemo-private/nemo-search-engine-simple.c:95:2: note: in expansion of macro ‘g_return_val_if_fail’
  g_return_val_if_fail (delimiter != NULL, NULL);
  ^~~~~~~~~~~~~~~~~~~~
../libnemo-private/nemo-search-engine-simple.c:101:4: warning: assignment discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
  s = remainder;
    ^
../libnemo-private/nemo-search-engine-simple.c:121:6: warning: assignment discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
    s = remainder;
      ^
[109/266] Compiling c object 'libnemo-...ate@sha/nemo-undo-signal-handlers.c.o'
../libnemo-private/nemo-undo-signal-handlers.c: In function ‘editable_register_edit_undo’:
../libnemo-private/nemo-undo-signal-handlers.c:165:9: warning: pointer targets in passing argument 2 of ‘gtk_editable_get_selection_bounds’ differ in signedness [-Wpointer-sign]
         &undo_data->selection_start,
         ^
In file included from /usr/include/gtk-3.0/gtk/gtkentry.h:37:0,
                 from /usr/include/gtk-3.0/gtk/gtktreeview.h:29,
                 from /usr/include/gtk-3.0/gtk/gtkcombobox.h:27,
                 from /usr/include/gtk-3.0/gtk/gtkappchooserbutton.h:29,
                 from /usr/include/gtk-3.0/gtk/gtk.h:42,
                 from ../libnemo-private/nemo-undo-signal-handlers.c:27:
/usr/include/gtk-3.0/gtk/gtkeditable.h:90:10: note: expected ‘gint * {aka int *}’ but argument is of type ‘guint * {aka unsigned int *}’
 gboolean gtk_editable_get_selection_bounds (GtkEditable *editable,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../libnemo-private/nemo-undo-signal-handlers.c:166:9: warning: pointer targets in passing argument 3 of ‘gtk_editable_get_selection_bounds’ differ in signedness [-Wpointer-sign]
         &undo_data->selection_end);
         ^
In file included from /usr/include/gtk-3.0/gtk/gtkentry.h:37:0,
                 from /usr/include/gtk-3.0/gtk/gtktreeview.h:29,
                 from /usr/include/gtk-3.0/gtk/gtkcombobox.h:27,
                 from /usr/include/gtk-3.0/gtk/gtkappchooserbutton.h:29,
                 from /usr/include/gtk-3.0/gtk/gtk.h:42,
                 from ../libnemo-private/nemo-undo-signal-handlers.c:27:
/usr/include/gtk-3.0/gtk/gtkeditable.h:90:10: note: expected ‘gint * {aka int *}’ but argument is of type ‘guint * {aka unsigned int *}’
 gboolean gtk_editable_get_selection_bounds (GtkEditable *editable,
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[116/266] Compiling c object 'libnemo-...private@sha/nemo-widget-menu-item.c.o'
../libnemo-private/nemo-widget-menu-item.c: In function ‘nemo_widget_menu_item_update’:
../libnemo-private/nemo-widget-menu-item.c:254:7: warning: implicit declaration of function ‘strcmp’ [-Wimplicit-function-declaration]
   if (strcmp (property_name, "widget-a") == 0)
       ^~~~~~
[125/266] Compiling c object 'src/nemo@exe/nemo-blank-desktop-window.c.o'
../src/nemo-blank-desktop-window.c:54:14: warning: ‘signals’ defined but not used [-Wunused-variable]
 static guint signals[LAST_SIGNAL] = { 0 };
              ^~~~~~~
[129/266] Linking target libnemo-private/libnemo-private.so
FAILED: libnemo-private/libnemo-private.so 
cc  -o libnemo-private/libnemo-private.so 'libnemo-private/nemo-private@sha/libnemo-private_nemo-generated.c.o' 'libnemo-private/nemo-private@sha/nemo-action-manager.c.o' 'libnemo-private/nemo-private@sha/nemo-action.c.o' 'libnemo-private/nemo-private@sha/nemo-bookmark.c.o' 'libnemo-private/nemo-private@sha/nemo-cell-renderer-disk.c.o' 'libnemo-private/nemo-private@sha/nemo-centered-placement-grid.c.o' 'libnemo-private/nemo-private@sha/nemo-clipboard-monitor.c.o' 'libnemo-private/nemo-private@sha/nemo-clipboard.c.o' 'libnemo-private/nemo-private@sha/nemo-column-chooser.c.o' 'libnemo-private/nemo-private@sha/nemo-column-utilities.c.o' 'libnemo-private/nemo-private@sha/nemo-context-menu-menu-item.c.o' 'libnemo-private/nemo-private@sha/nemo-dbus-manager.c.o' 'libnemo-private/nemo-private@sha/nemo-debug.c.o' 'libnemo-private/nemo-private@sha/nemo-default-file-icon.c.o' 'libnemo-private/nemo-private@sha/nemo-desktop-directory-file.c.o' 'libnemo-private/nemo-private@sha/nemo-desktop-directory.c.o' 'libnemo-private/nemo-private@sha/nemo-desktop-icon-file.c.o' 'libnemo-private/nemo-private@sha/nemo-desktop-link-monitor.c.o' 'libnemo-private/nemo-private@sha/nemo-desktop-link.c.o' 'libnemo-private/nemo-private@sha/nemo-desktop-metadata.c.o' 'libnemo-private/nemo-private@sha/nemo-desktop-utils.c.o' 'libnemo-private/nemo-private@sha/nemo-directory-async.c.o' 'libnemo-private/nemo-private@sha/nemo-directory.c.o' 'libnemo-private/nemo-private@sha/nemo-dnd.c.o' 'libnemo-private/nemo-private@sha/nemo-entry.c.o' 'libnemo-private/nemo-private@sha/nemo-file-changes-queue.c.o' 'libnemo-private/nemo-private@sha/nemo-file-conflict-dialog.c.o' 'libnemo-private/nemo-private@sha/nemo-file-dnd.c.o' 'libnemo-private/nemo-private@sha/nemo-file-operations.c.o' 'libnemo-private/nemo-private@sha/nemo-file-queue.c.o' 'libnemo-private/nemo-private@sha/nemo-file-undo-manager.c.o' 'libnemo-private/nemo-private@sha/nemo-file-undo-operations.c.o' 'libnemo-private/nemo-private@sha/nemo-file-utilities.c.o' 'libnemo-private/nemo-private@sha/nemo-file.c.o' 'libnemo-private/nemo-private@sha/nemo-global-preferences.c.o' 'libnemo-private/nemo-private@sha/nemo-icon-canvas-item.c.o' 'libnemo-private/nemo-private@sha/nemo-icon-container.c.o' 'libnemo-private/nemo-private@sha/nemo-icon-dnd.c.o' 'libnemo-private/nemo-private@sha/nemo-icon-info.c.o' 'libnemo-private/nemo-private@sha/nemo-job-queue.c.o' 'libnemo-private/nemo-private@sha/nemo-lib-self-check-functions.c.o' 'libnemo-private/nemo-private@sha/nemo-link.c.o' 'libnemo-private/nemo-private@sha/nemo-merged-directory.c.o' 'libnemo-private/nemo-private@sha/nemo-metadata.c.o' 'libnemo-private/nemo-private@sha/nemo-mime-application-chooser.c.o' 'libnemo-private/nemo-private@sha/nemo-module.c.o' 'libnemo-private/nemo-private@sha/nemo-monitor.c.o' 'libnemo-private/nemo-private@sha/nemo-placement-grid.c.o' 'libnemo-private/nemo-private@sha/nemo-places-tree-view.c.o' 'libnemo-private/nemo-private@sha/nemo-program-choosing.c.o' 'libnemo-private/nemo-private@sha/nemo-progress-info-manager.c.o' 'libnemo-private/nemo-private@sha/nemo-progress-info.c.o' 'libnemo-private/nemo-private@sha/nemo-query.c.o' 'libnemo-private/nemo-private@sha/nemo-recent.c.o' 'libnemo-private/nemo-private@sha/nemo-saved-search-file.c.o' 'libnemo-private/nemo-private@sha/nemo-search-directory-file.c.o' 'libnemo-private/nemo-private@sha/nemo-search-directory.c.o' 'libnemo-private/nemo-private@sha/nemo-search-engine-simple.c.o' 'libnemo-private/nemo-private@sha/nemo-search-engine.c.o' 'libnemo-private/nemo-private@sha/nemo-selection-canvas-item.c.o' 'libnemo-private/nemo-private@sha/nemo-separator-action.c.o' 'libnemo-private/nemo-private@sha/nemo-signaller.c.o' 'libnemo-private/nemo-private@sha/nemo-thumbnails.c.o' 'libnemo-private/nemo-private@sha/nemo-trash-monitor.c.o' 'libnemo-private/nemo-private@sha/nemo-tree-view-drag-dest.c.o' 'libnemo-private/nemo-private@sha/nemo-ui-utilities.c.o' 'libnemo-private/nemo-private@sha/nemo-undo-manager.c.o' 'libnemo-private/nemo-private@sha/nemo-undo-signal-handlers.c.o' 'libnemo-private/nemo-private@sha/nemo-undo-transaction.c.o' 'libnemo-private/nemo-private@sha/nemo-undo.c.o' 'libnemo-private/nemo-private@sha/nemo-vfs-directory.c.o' 'libnemo-private/nemo-private@sha/nemo-vfs-file.c.o' 'libnemo-private/nemo-private@sha/nemo-widget-action.c.o' 'libnemo-private/nemo-private@sha/nemo-widget-menu-item.c.o' '-Wl,--no-undefined' '-Wl,--as-needed' '-shared' '-fPIC' '-Wl,-soname,libnemo-private.so' 'eel/libeel.a' 'libnemo-extension/libnemo-extension.so.1.4.0' 'eel/libeel.a' 'eel/libeel.a' '-lcinnamon-desktop' '-lgtk-3' '-lgdk-3' '-lpangocairo-1.0' '-lpango-1.0' '-latk-1.0' '-lcairo-gobject' '-lcairo' '-lgdk_pixbuf-2.0' '-lgio-2.0' '-lgobject-2.0' '-lglib-2.0' '-lgtk-3' '-lgdk-3' '-lpangocairo-1.0' '-lpango-1.0' '-latk-1.0' '-lcairo-gobject' '-lcairo' '-lgdk_pixbuf-2.0' '-lgio-2.0' '-lgobject-2.0' '-lglib-2.0' '-lxml2' '-lgailutil-3' '-lgtk-3' '-lgdk-3' '-lpangocairo-1.0' '-lpango-1.0' '-latk-1.0' '-lcairo-gobject' '-lcairo' '-lgdk_pixbuf-2.0' '-lgio-2.0' '-lgobject-2.0' '-lglib-2.0' '-lgailutil-3' '-lgtk-3' '-lgdk-3' '-lpangocairo-1.0' '-lpango-1.0' '-latk-1.0' '-lcairo-gobject' '-lcairo' '-lgdk_pixbuf-2.0' '-lgio-2.0' '-lgobject-2.0' '-lglib-2.0' '-lglib-2.0' '-lgtk-3' '-lgdk-3' '-lpangocairo-1.0' '-lpango-1.0' '-latk-1.0' '-lcairo-gobject' '-lcairo' '-lgdk_pixbuf-2.0' '-lgio-2.0' '-lgobject-2.0' '-lglib-2.0' '-lexif' '-lxml2' '-lm' '-lglib-2.0' '-lgtk-3' '-lgdk-3' '-lpangocairo-1.0' '-lpango-1.0' '-latk-1.0' '-lcairo-gobject' '-lcairo' '-lgdk_pixbuf-2.0' '-lgio-2.0' '-lgobject-2.0' '-lglib-2.0' '-lgtk-3' '-lgdk-3' '-lpangocairo-1.0' '-lpango-1.0' '-latk-1.0' '-lcairo-gobject' '-lcairo' '-lgdk_pixbuf-2.0' '-lgio-2.0' '-lgobject-2.0' '-lglib-2.0' '-lxml2' '-lgailutil-3' '-lgtk-3' '-lgdk-3' '-lpangocairo-1.0' '-lpango-1.0' '-latk-1.0' '-lcairo-gobject' '-lcairo' '-lgdk_pixbuf-2.0' '-lgio-2.0' '-lgobject-2.0' '-lglib-2.0' '-lX11' '-lgtk-3' '-lgdk-3' '-lpangocairo-1.0' '-lpango-1.0' '-latk-1.0' '-lcairo-gobject' '-lcairo' '-lgdk_pixbuf-2.0' '-lgio-2.0' '-lgobject-2.0' '-lglib-2.0' '-lxml2' '-lgailutil-3' '-lgtk-3' '-lgdk-3' '-lpangocairo-1.0' '-lpango-1.0' '-latk-1.0' '-lcairo-gobject' '-lcairo' '-lgdk_pixbuf-2.0' '-lgio-2.0' '-lgobject-2.0' '-lglib-2.0' '-lgtk-3' '-lgdk-3' '-lpangocairo-1.0' '-lpango-1.0' '-latk-1.0' '-lcairo-gobject' '-lcairo' '-lgdk_pixbuf-2.0' '-lgio-2.0' '-lgobject-2.0' '-lglib-2.0' '-lxml2' '-lgailutil-3' '-lgtk-3' '-lgdk-3' '-lpangocairo-1.0' '-lpango-1.0' '-latk-1.0' '-lcairo-gobject' '-lcairo' '-lgdk_pixbuf-2.0' '-lgio-2.0' '-lgobject-2.0' '-lglib-2.0' '-Wl,-rpath,/home/leigh/Downloads/nemo-MesonPort/build/libnemo-private:/home/leigh/Downloads/nemo-MesonPort/build/libnemo-extension'  
libnemo-private/nemo-private@sha/nemo-file.c.o: In function `nemo_file_get_selinux_context':
/home/leigh/Downloads/nemo-MesonPort/build/../libnemo-private/nemo-file.c:5217: undefined reference to `selinux_raw_to_trans_context'
libnemo-private/nemo-private@sha/nemo-file.c.o: In function `nemo_file_get_selinux_context':
nemo-file.c:(.text+0xb7d9): undefined reference to `freecon'
libnemo-private/nemo-private@sha/nemo-module.c.o: In function `nemo_module_load':
/home/leigh/Downloads/nemo-MesonPort/build/../libnemo-private/nemo-module.c:42: undefined reference to `g_module_open'
libnemo-private/nemo-private@sha/nemo-module.c.o: In function `nemo_module_load':
nemo-module.c:(.text+0x13b): undefined reference to `g_module_error'
nemo-module.c:(.text+0x182): undefined reference to `g_module_symbol'
nemo-module.c:(.text+0x1a5): undefined reference to `g_module_symbol'
nemo-module.c:(.text+0x1c8): undefined reference to `g_module_symbol'
nemo-module.c:(.text+0x1d1): undefined reference to `g_module_error'
nemo-module.c:(.text+0x1ff): undefined reference to `g_module_close'
libnemo-private/nemo-private@sha/nemo-module.c.o: In function `nemo_module_unload':
nemo-module.c:(.text+0x25f): undefined reference to `g_module_close'
libnemo-private/nemo-private@sha/nemo-thumbnails.c.o: In function `thumbnail_thread_starter_cb':
/home/leigh/Downloads/nemo-MesonPort/build/../libnemo-private/nemo-thumbnails.c:182: undefined reference to `pthread_create'
collect2: error: ld returned 1 exit status
[130/266] Compiling c object 'src/nemo@exe/nemo-desktop-icon-grid-view.c.o'
../src/nemo-desktop-icon-grid-view.c: In function ‘nemo_desktop_icon_grid_view_class_init’:
../src/nemo-desktop-icon-grid-view.c:375:38: warning: assignment from incompatible pointer type [-Wincompatible-pointer-types]
  G_OBJECT_CLASS (class)->constructed = nemo_desktop_icon_grid_view_constructed;
                                      ^
../src/nemo-desktop-icon-grid-view.c: In function ‘set_sort_type’:
../src/nemo-desktop-icon-grid-view.c:760:54: warning: passing argument 1 of ‘nemo_view_get_selection’ from incompatible pointer type [-Wincompatible-pointer-types]
                 selection = nemo_view_get_selection (view);
                                                      ^~~~
In file included from ../src/nemo-icon-view.h:29:0,
                 from ../src/nemo-desktop-icon-grid-view.h:28,
                 from ../src/nemo-desktop-icon-grid-view.c:29:
../src/nemo-view.h:390:19: note: expected ‘NemoView * {aka struct NemoView *}’ but argument is of type ‘NemoDesktopIconGridView * {aka struct <anonymous> *}’
 GList *           nemo_view_get_selection              (NemoView      *view);
                   ^~~~~~~~~~~~~~~~~~~~~~~
../src/nemo-desktop-icon-grid-view.c: In function ‘real_update_menus’:
../src/nemo-desktop-icon-grid-view.c:934:19: warning: unused variable ‘zoom_level’ [-Wunused-variable]
     NemoZoomLevel zoom_level;
                   ^~~~~~~~~~
[134/266] Compiling c object 'src/nemo@exe/nemo-pathbar.c.o'
../src/nemo-pathbar.c: In function ‘nemo_path_bar_update_path’:
../src/nemo-pathbar.c:2060:5: warning: statement with no effect [-Wunused-value]
     last_directory;
     ^~~~~~~~~~~~~~
[138/266] Compiling c object 'src/nemo@exe/nemo-error-reporting.c.o'
ninja: build stopped: subcommand failed.
```

Salamandar commented:
Fixed. Thanks for the feedback, I know there probably are some problems (#defines not defined maybe).
And btw, localization is not handled as I don't understand how it's handled with Autotools today. There are no .pot files… ?

clefebvre commented:
hi @Salamandar,

l10n is handled separately, it's not part of the build system.

makepot produces a pot file we send towards Launchpad.
Launchpad produces po files we feed into cinnamon-translations.

clefebvre commented:
Autotools can take care of it, but it's not convenient at all in the scope of Cinnamon. It relies on hardcoded LINGUAS, POTFILES.in and it's something we'd need to do for each Cinnamon component individually even though we don't ship them in isolation.

Salamandar commented:
@clefebvre Ok, I can see that.
Does that mean that, with Autotools or Meson, the nemo build system will never see the translations ? If yes I'll remove the consequent code.

clefebvre commented:
Yes, the build shouldn't worry about l10n at all.

Salamandar commented:
It may be easy, with Nemo ported, to have other linuxmint projects ported as well.
Can you confirm the port is correct ?

clefebvre commented:
Hi @Salamandar, I'll have to catch up on Meson. It looks really exciting but it's quite new to us.

We'll also need to decide on whether or not we want to fully switch to it. Maintaining two build systems might not be a good idea. If meson does everything we need, we might want to get rid of autotools.

I can't look into it just yet though. We're at the end of our development cycle and getting ready to release. I've marked your PR for 3.6.

clefebvre commented:
Hi @Salamandar,

Can you continue to edit this PR?

**Dep on xapps**

A new dependency was added.

In buid.meson:

`xapp    = dependency('xapp')`

In src/build.meson:

```
nemo_deps = [ cinnamon, exempi, gail, glib, gtk, libexif, libnotify, libxml, math, x11,
    egg, nemo_extension, nemo_private, xapp ]
```

clefebvre commented:
I'm getting a bit further but things get installed in the wrong places...

Here's my debian/rules:

```
#!/usr/bin/make -f

TYPELIBDIR=$(shell pkg-config gobject-introspection-1.0 --variable libdir | sed -e 's/.//')

CONFIGURE_EXTRA_FLAGS = --prefix=/usr --libexecdir=/usr/lib/nemo
			# 			--disable-update-mimedb \
			# --enable-gtk-doc \
			# --enable-selinux=no \
			# --enable-compile-warnings=yes \
			# --enable-silent-rules

export LDFLAGS+=-Wl,-z,defs -Wl,-O1 -Wl,--as-needed


%:
	dh $@ --parallel --with gir

override_dh_auto_clean:
	rm -rf build

override_dh_auto_configure:
	mkdir -p build
	LIBEXTDIR=/usr/lib/nemo meson $(CONFIGURE_EXTRA_FLAGS) ./ build/

override_dh_auto_build:
	cd build && ninja

override_dh_auto_test:

override_dh_strip:
	dh_strip --dbg-package=nemo-dbg

override_dh_install:
	sed 's@TYPELIBDIR@${TYPELIBDIR}@' debian/gir1.2-nemo-3.0.install.in > debian/gir1.2-nemo-3.0.install
	cd build && DESTDIR=${CURDIR}/debian/tmp ninja install
	dh_install --list-missing

override_dh_makeshlibs:
	dh_makeshlibs -Xusr/lib/nemo/extensions-3.0/

override_dh_installmime:
	dh_installmime -pnemo-data

```

clefebvre commented:
I had to change libExecPath = get_option('libexecdir') in meson.build and a few install_dir here and there in src and other subdirs...

I'm still getting stuff in the wrong dirs though. dh_install lists what's in the wrong place:

dh_install --list-missing
dh_install: Compatibility levels before 9 are deprecated (level 8 in use)
dh_install: Cannot find (any matches for) "usr/lib/libnemo-extension*.so.*" (tried in "." and "debian/tmp")
dh_install: libnemo-extension1 missing files: usr/lib/libnemo-extension*.so.*
dh_install: Cannot find (any matches for) "usr/lib/libnemo-extension*.so" (tried in "." and "debian/tmp")
dh_install: libnemo-extension-dev missing files: usr/lib/libnemo-extension*.so
dh_install: Cannot find (any matches for) "usr/lib/pkgconfig" (tried in "." and "debian/tmp")
dh_install: libnemo-extension-dev missing files: usr/lib/pkgconfig
dh_install: usr/lib/libnemo-extension.pc exists in debian/tmp but is not installed to anywhere
dh_install: usr/lib/x86_64-linux-gnu/libnemo-extension.so exists in debian/tmp but is not installed to anywhere
dh_install: usr/lib/x86_64-linux-gnu/libnemo-private.so exists in debian/tmp but is not installed to anywhere
dh_install: usr/lib/x86_64-linux-gnu/libnemo-extension.so.1.4.0 exists in debian/tmp but is not installed to anywhere
dh_install: usr/lib/x86_64-linux-gnu/libnemo-extension.so.1 exists in debian/tmp but is not installed to anywhere
dh_install: missing files, aborting

Salamandar commented:
Hi @clefebvre.
No problem to work on it.
I was expecting some packaging "inconsistencies" with the Autotools package, mostly on the Debian pkg. I'll (try to) fix them tonight or tomorrow.

Salamandar commented:
Actually your 

    CONFIGURE_EXTRA_FLAGS = --prefix=/usr --libexecdir=/usr/lib/nemo

should be

    CONFIGURE_EXTRA_FLAGS = --prefix=/usr --libdir=/usr/lib --libexecdir=/usr/lib/nemo

Don't ask me why, but it seems that the default LibDir is /usr/lib/x86_64-linux-gnu… But on Debian only, not ArchLinux.

Fantu commented:
if can be useful debhelper 10.3 added meson+ninja build support:
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=795253

Salamandar commented:
@Fantu Seems cool. But in how many Debian/Mint versions will it be available ? :p
Joke apart, that'd simplify the process if it's available right away.

clefebvre commented:
We've got debhelper 10.2 in both Xenial and Stretch.

clefebvre commented:
I forgot something else... the files/ directory is designed to place files "as is" on the filesystem. So anything in files/A/B/C should end up in /A/B/C on the filesystem once the app is installed.

With autotools we had a Makefile.am in there which dynamically included anything we shoved in there. Can the same be done with Meson? so that we can continue to place files in files/ and not have to edit meson.build each time we do that?

clefebvre commented:
So I'm getting a bit further...

I changed the typelib dir in debian/rules:

`TYPELIBDIR=usr/lib`

It might create issues with the gir further on, but we can look into that later.

In debian/rules, I also had to comment:
```
			# --disable-update-mimedb \
			# --enable-gtk-doc \
			# --enable-selinux=no \
			# --enable-compile-warnings=yes \
			# --enable-silent-rules
```

After this is done, I'm getting a successful build and the deb packages are produced.

At runtime, if I run nemo though:

> nemo: error while loading shared libraries: libnemo-private.so: cannot open shared object file: No such file or directory

I'm testing this in Mint 18.2, which package base is Ubuntu 16.04 Xenial. The version of meson is 0.40.1.

Salamandar commented:
@clefebvre So it seems Github did not want my answer.
> With autotools we had a Makefile.am in there which dynamically included anything we shoved in there.
By design, Meson does not provide any "file wildcards". The reason for that is "keep the build system simple". Not that I agree on that.
The solution given in the Meson doc is : call a shell script that will do the job. That may be the best solution ;)

clefebvre commented:
As long as it works :)

Salamandar commented on src/nemo-empty-view.c:
```diff
@@ -55,7 +55,7 @@ nemo_empty_view_add_file (NemoView *view, NemoFile *file, NemoDirectory *directo
 
 	g_timer_start (timer);
-	icon = nemo_file_get_icon_pixbuf (file, nemo_get_icon_size_for_zoom_level (NEMO_ZOOM_LEVEL_STANDARD), TRUE, 0);
+	icon = nemo_file_get_icon_pixbuf (file, nemo_get_icon_size_for_zoom_level (NEMO_ZOOM_LEVEL_STANDARD), TRUE, 0, NEMO_FILE_ICON_FLAGS_NONE);
```
I had to do that to have a successful build. Dunno why.

Salamandar commented on src/nemo-desktop-application.c:
```diff
@@ -278,7 +278,7 @@ nemo_desktop_application_local_command_line (GApplication *application,
 
     if (version) {
-        g_print ("nemo-desktop " PACKAGE_VERSION "\n");
+        g_print ("nemo-desktop " VERSION "\n");
```
Unified #defined variables

Salamandar commented on src/nemo-main-application.c:
```diff
@@ -543,7 +543,7 @@ nemo_main_application_local_command_line (GApplication *application,
 
 	if (version) {
-		g_print ("nemo " PACKAGE_VERSION "\n");
+		g_print ("nemo " VERSION "\n");
```
Unified #defined variables

mtwebster commented:
Hi, do you think it'd be possible for you to rebase this on current master?  I've got a couple fixes to add and I'd like to get this merged.  If not, I could probably manage to rebase it all, but it'd lose its association with this PR (though your commits would still be there)

Thanks!

Salamandar commented:
I'm doing this right now. As a rebase may be impossible to do, i'm doing a simple branch merge.

Salamandar commented:
Merge and some cleanup done. Now, Meson is just missing from the CI docker images. ^^

mtwebster commented:
Do you mind if I squash this before I merge?  I'd rather have just one or two commits delineating our switch to meson.  I managed to do a real rebase so it'll all merge on top of master now.  I've got a followup commit for it that fixes various stuff for our older mint base, along with cleanup and some re-organization (made much simpler with meson!)

(working branch: https://github.com/mtwebster/nemo/tree/MesonPort-rebase)

Thanks alot.

Salamandar commented:
I reviewed your changes. Basically, i'm okay with those changes apart from those i commented.
For the post-install scripts, I understand the motivation but I'm afraid that short-circuits the package manager behaviour.
For the files installation, why forcing the prefix ? Some distros will use separate prefixes for some softwares, and that forbids an installation in /home/my_user/my_soft (some companies don't leave root permissions). And you can forget about MingW.

Salamandar commented:
I squashed most of the commits, but I split the sources (*.c) changes from the build-system changes.
The commit https://github.com/linuxmint/nemo/pull/1493/commits/cc0a066f5e1a87bb7b5872919258bb772672abf8 only has Git merge tags. I don't thinks it's normal ;)

mtwebster commented:
Please don't re-run makepot right now, that's something done by clem nearer to release time.

mtwebster commented:
I merged all of this except the "Fix gdbus code-gen" commit and the gettext/python changes commit.

The codegen one was very mixed up, some of the changes were unrelated, and already done in my main commit.  The gettext is fine, I just don't want to update nemo.pot, please open another PR for that one cleaned up.  Thanks alot

Salamandar commented:
Great :smile:
The "fix gdbus" commit was just part of your big commit, not actually my changes as you just saw.
